packages/ipfs-core/src/components/dag/export.js

```diff
@@ -0,0 +1,133 @@
+/** @type {number[]} */
+const NO_LINKS_CODECS = [
+  dagCbor.code, // CBOR
+  raw.code, // raw
+  json.code // JSON
+]
+
+/**
+ * @typedef {import('../../types').Preload} Preload
+ * @typedef {import('ipfs-repo').IPFSRepo} IPFSRepo
+ * @typedef {import('@ipld/car/api').BlockWriter} BlockWriter
+ * @typedef {import('ipfs-core-types/src/utils').AbortOptions} AbortOptions
+ */
+
+/**
+ * @param {Object} config
+ * @param {IPFSRepo} config.repo
+ * @param {Preload} config.preload
+ * @param {import('ipfs-core-utils/src/multicodecs')} config.codecs
+ */
+module.exports = ({ repo, preload, codecs }) => {
+  /**
+   * @type {import('ipfs-core-types/src/dag').API["export"]}
+   */
+  async function * dagExport (root, options = {}) {
+    if (options.preload !== false) {
+      preload(root)
+    }
+
+    const cid = CID.asCID(root)
+    if (!cid) {
+      throw new Error(`Unexpected error converting CID type: ${root}`)
+    }
+
+    log(`Exporting ${cid} as car`)
+    const { writer, out } = await CarWriter.create([cid])
+
+    // we need to write with one async channel and send the CarWriter output
+    // with another to the caller, but if the write causes an error we capture
+    // that and make sure it gets propagated
+    /** @type {Error|null} */
+    let err = null
+    ;(async () => {
+      try {
+        await traverseWrite(
+          repo,
+          { signal: options.signal, timeout: options.timeout },
+          cid,
+          writer,
+          codecs)
+        writer.close()
+      } catch (e) {
+        err = e
+      }
+    })()
+
+    for await (const chunk of out) {
+      if (err) {
+        break
+      }
+      yield chunk
+    }
+    if (err) {
+      throw err
+    }
+  }
+
+  return withTimeoutOption(dagExport)
+}
+
+/**
+ * @param {IPFSRepo} repo
+ * @param {AbortOptions} options
+ * @param {CID} cid
+ * @param {BlockWriter} writer
+ * @param {import('ipfs-core-utils/src/multicodecs')} codecs
+ * @param {Set<string>} seen
+ * @returns {Promise<void>}
+ */
+async function traverseWrite (repo, options, cid, writer, codecs, seen = new Set()) {
+  const b58Cid = cid.toString(base58btc)
+  if (seen.has(b58Cid)) {
+    return
+  }
+
+  const block = await getBlock(repo, options, cid, codecs)
+
+  log(`Adding block ${cid} to car`)
+  await writer.put(block)
+  seen.add(b58Cid)
+
+  // recursive traversal of all links
+  for (const link of block.links) {
+    await traverseWrite(repo, options, link, writer, codecs, seen)
+  }
+}
+
+/**
+ * @param {IPFSRepo} repo
+ * @param {AbortOptions} options
+ * @param {CID} cid
+ * @param {import('ipfs-core-utils/src/multicodecs')} codecs
+ * @returns {Promise<{cid:CID, bytes:Uint8Array, links:CID[]}>}
+ */
+async function getBlock (repo, options, cid, codecs) {
+  const bytes = await repo.blocks.get(cid, options)
+
+  /** @type {CID[]} */
+  let links = []
+  const codec = await codecs.getCodec(cid.code)
+
+  if (codec) {
+    const block = Block.createUnsafe({ bytes, cid, codec })
```

Comment:
One difference we might find with Go is that they're going to be obsessively checking CIDs are correct for the bytes whenever they're loaded, at least in cases like this. The `Unsafe` is giving us a shortcut, basically saying we don't care if it doesn't match, and in fact we probably trust the underlying source of blocks to have already done that check. The new go-ipld-prime integration work is providing some opportunities for "trusted" data sources that skip these checks too so they may get slight perf improvements.

Reply:
In this case I don't think we need to as the blockstore creates the path to the block from the multihash extracted from the CID.  I suppose you could mess around with the fs/level backing store to change the contents of the buffer stored at the relevant path but I think solving that problem is outside the scope of dag export.

Re-hashing/verifying blocks on every export would be quite expensive, definitely something we should put behind a flag if people need those kind of guarantees.

go-ipfs has a `ipfs repo verify` command which I guess would load every block in the repo and rehash it to ensure we've got the right bytes, there's no equivalent in js land yet.

Reply:
related: the old `Datastore.HashOnRead` option that is `false` by default, due to the perf hit and "i trust my datastore" assumption https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#datastorehashonread

Reply:
Go will be getting the same treatment when the ipld-prime work makes its way through: https://github.com/ipfs/go-fetcher/blob/64b1f390e7ae13d96494f15a10e07527369a521d/impl/blockservice/fetcher.go#L45

I'm pretty sure (but far from certain) that any time you interact with an IPLD `Node` at the moment it does the whole-hog of verification and you can't bypass it. In fact there's a lot of places where you can't even avoid decoding the bytes even if you just want the bytes. Things should get a little more sophisticated as ipld-prime moves through.